python/cudf_polars/cudf_polars/experimental/groupby.py

```diff
@@ -0,0 +1,210 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Parallel GroupBy Logic."""
+
+from __future__ import annotations
+
+import operator
+from functools import reduce
+from typing import TYPE_CHECKING, Any
+
+import pylibcudf as plc
+
+from cudf_polars.dsl.expr import Agg, BinOp, Cast, Col, Len, NamedExpr
+from cudf_polars.dsl.ir import GroupBy, Select
+from cudf_polars.experimental.base import PartitionInfo, _concat, get_key_name
+from cudf_polars.experimental.dispatch import generate_ir_tasks, lower_ir_node
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.dsl.expr import Expr
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.parallel import LowerIRTransformer
+
+
+class GroupByTree(GroupBy):
+    """Groupby tree-reduction operation."""
```

Comment:
*design discussion*: This node is "just" a `GroupBy` whose only job is to indicate to the lowering system that the multiple partitions of the "child" should be lowered via tree reduction. I think that's OK, but _perhaps_ we can get away with handling this just using `GroupBy` and using the `partition_count` of a groupby node to determine what to do.

Reply:
Yes, that's fine. It was a bit easier to think through the problem with an explicit sub-class in hand, but we can definitely use partitioning information instead.